daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Transaction.scala

```diff
@@ -613,4 +604,176 @@ object Transaction {
       fa: FailedAuthorization,
   ) extends TransactionError
 
+  /** Whether `other` is the result of reinterpreting this transaction.
+    *
+    * @note This function is asymmetric.
+    */
+  def isReplayedBy[Nid, Cid, Val](
+      recorded: GenTransaction[Nid, Cid, Val],
+      replayed: GenTransaction[Nid, Cid, Val],
+  )(implicit ECid: Equal[Cid], EVal: Equal[Val]): Either[ReplayMismatch[Nid, Cid, Val], Unit] = {
+    import scalaz.syntax.equal._
+    import scalaz.std.option._
+
+    type Exe = Node.NodeExercises[Nid, Cid, Val]
+
+    def fail(recordedMismatch: Option[Nid], replayMismatch: Option[Nid]) =
+      Left(ReplayMismatch(recorded, recordedMismatch, replayed, replayMismatch))
+
+    @tailrec
+    def loop(
+        nids1: Stream[Nid],
+        nids2: Stream[Nid],
+        stack: List[(Nid, Exe, Stream[Nid], Nid, Exe, Stream[Nid])] = List.empty,
+    ): Either[ReplayMismatch[Nid, Cid, Val], Unit] =
+      (nids1, nids2) match {
+        case (nid1 #:: rest1, nid2 #:: rest2) =>
+          (recorded.nodes(nid1), replayed.nodes(nid2)) match {
+            case (
+                Node.NodeCreate(
+                  coid1,
+                  coinst1,
+                  optLocation1 @ _,
+                  signatories1,
+                  stakeholders1,
+                  key1
+                ),
+                Node.NodeCreate(
+                  coid2,
+                  coinst2,
+                  optLocation2 @ _,
+                  signatories2,
+                  stakeholders2,
+                  key2
+                ))
+                if coid1 === coid2 &&
+                  coinst1 === coinst2 &&
+                  signatories1 == signatories2 &&
+                  stakeholders1 == stakeholders2 &&
+                  key1 === key2 =>
+              loop(rest1, rest2, stack)
+            case (
+                Node.NodeFetch(
+                  coid1,
+                  templateId1,
+                  optLocation1 @ _,
+                  actingParties1,
+                  signatories1,
+                  stakeholders1,
+                  key1,
+                  byKey1 @ _,
+                ),
+                Node.NodeFetch(
+                  coid2,
+                  templateId2,
+                  optLocation2 @ _,
+                  actingParties2,
+                  signatories2,
+                  stakeholders2,
+                  key2,
+                  byKey2 @ _,
+                ))
+                if coid1 === coid2 &&
+                  templateId1 == templateId2 &&
+                  (actingParties1.isEmpty || actingParties1 == actingParties2) &&
```

Comment:
This condition looks asymmetric. Is that intended? Same question below for `key`.

Reply:
Yes, this is for backward compatibility. 
The first transaction is the `recorded` one, (i.e the transaction generated by another node), while the second transaction is the `replayed` one.
In transaction version < 5, `actingParties1` is `None`, while in recent engines it is always `Some` of something.

Reply:
That makes sense. Thanks for the explanation. Could you please a comment somewhere towards to top of this function explaining that it is not as symmetric as it (mostly) looks like for these reasons?

Reply:
sure.